src/runtime_src/hip/core/module.h

```diff
@@ -48,25 +74,47 @@ class module
   {
     return function_cache.get(handle);
   }
+
+  const xrt::hw_context&
+  get_hw_context()
+  {
+    return m_hw_ctx;
+  }
+};
+
+class module_elf : public module
+{
+  xrt::elf m_elf;
+  xrt::module m_module;
+  module_xclbin* m_xclbin_module;
+
+public:
+  module_elf(module_xclbin* xclbin_module, const std::string& file_name);
+
+  module_xclbin*
+  get_xclbin_module() const { return m_xclbin_module; }
+
+  const xrt::module&
+  get_xrt_module() const { return m_module; }
 };
 
 class function
 {
-  module* m_module;
+  module_xclbin* m_module = nullptr;
   std::string m_func_name;
-  xrt::kernel m_kernel;
+  xrt::ext::kernel m_kernel;
```

Comment:
```suggestion
  xrt::kernel m_kernel;
```
xrt::ext::kernel is just for construction, but assignment should be to xrt::kernel.  You can remove the default xrt::ext::kernel ctor.